Modules/main.c

```diff
@@ -363,6 +363,7 @@ pymain_init(const _PyArgv *args, PyInterpreterState **interp_p)
 
     err = config_read_write(config, args, preconfig);
     if (_Py_INIT_FAILED(err)) {
+        _PyRuntimeState_Fini(&_PyRuntime);
```

Comment:
That's not the right place to cleanup things. I redesigned deeply Py_Main() to replace exit() call in mutliple layers of function calls, far from Py_Main(), with _PyInitError which allows to give back the control flow to Py_Main(). I did that to be able to cleanup Python properly.